src/lib/y2storage/encryption_method/tpm_fde.rb

```diff
@@ -0,0 +1,151 @@
+    # if this method is used, it must be used at least for the root (/) filesystem and only
+    # additionally for some other devices.
+    class TpmFde < Base
+      def initialize
+        textdomain "storage"
+
+        super(:tpm_fde, _("TPM-Based Full Disk Encrytion"))
+      end
+
+      # @see Base#used_for?
+      #
+      # @todo Not sure if this would be possible at the end, since the exact way to setup the
+      # system using fde-tools is still changing too often. In any case, having a precise result
+      # for this method will only be relevant when implementing support for creating encrypted
+      # devices in an installed system. During installation returning always false is perfectly
+      # correct.
+      #
+      # @return [Boolean] false
+      def used_for?(_encryption)
+        # One candidate criteria (still waiting for some conversations with fde-toold developers)
+        # could be:
+        # encryption.type.is?(:luks2) && key_file == EncryptionProcesses::TpmFdeTools.key_file_name
+        false
+      end
+
+      # @see Base#available?
+      #
+      # In this initial implementation this always returns false because there important
+      # limitations to use this in (Auto)YaST:
+      #
+      # - The current version of the inst-sys cannot talk to the TPM
+      # - There is still no corresponding UI in the Expert Partitioner
+      # - Some mechanism to ensure consistency (eg. checking all devices use the same recovery
+      #   password) need to be introduced
+      # - The current implementation of the encryption method only covers system installation
+      #   (with no support to add a new encrypted device to a system already using fde-tools)
+      #
+      # So far, the encryption method is implemented to be used by Agama (which doesn't honor
+      # the {#available?} method.
+      #
+      # @return [Boolean] false
+      def available?
+        false
+      end
+
+      # Whether both the target system and the product being installed meet the requisites
+      # to setup devices using this encryption method.
+      #
+      # The encryption method must be used at least for the root filesystem (eg. is not possible to
+      # use it for /var but not for /), but that can't hardly be controlled here. A separate
+      # validation that considers the whole devicegraph is needed.
+      #
+      # @return [Boolean]
+      def possible?
+        tpm_system? && tpm_product?
+      end
+
+      # Creates an encryption device for the given block device
+      #
+      # @param blk_device [Y2Storage::BlkDevice]
+      # @param dm_name [String]
+      #
+      # @return [Y2Storage::Encryption]
+      def create_device(blk_device, dm_name, label: "")
+        encryption_process.create_device(blk_device, dm_name, label: label)
+      end
+
+      # Special method used in the tests to reset all memoizations
+      def reset
```

Comment:
NP: Is this strictly needed? Having it only because tests is not good. Maybe we can use some metaprograming in tests to set `@tpm_present` instance variable.

Reply:
I did it that way because that's the approach we followed with the memoization of available packages at `StorageFeatures` (a code I had to touch recently). But I'm fine with changing that.

Reply:
I realized this is actually not needed because the tests create a new instance of the encryption method on each example. So there is no caching affecting the tests.